### PR TITLE
Add GitHub dispatch provider

### PR DESCRIPTION
**Pull request by alekspog**
Add support for GitHub repository dispatch provider.

Ref: https://github.com/fluxcd/flux2/discussions/2122
Fix: https://github.com/fluxcd/notification-controller/issues/329

Example of the notification message payload in the GitHub workflow

Screenshot from 2022-04-22 10-07-16

### CODE REVIEW

File: docs/spec/v1beta1/provider.md [hunk 1]
```diff
@@ -730,3 +732,134 @@ kubectl create secret generic webhook-url \
...
+
+```yaml
+name: test-github-dispatch-provider
+on:
+  repository_dispatch:
+    types:
+      - "*app1-staging*"
+```
+
+If the information from `event_type` is not enough to get the necessary information about the configured resources, you can get more information from `client_payload`. For example, you can get the message from the payload in the workfow job step:
```
Comment (stefanprodan) [inline docs/spec/v1beta1/provider.md:818]: Can you please change this to show how `summary` can be used.

Comment (alekspog) [reply docs/spec/v1beta1/provider.md]: I have modified the examples in relation to our discussion in slack. The `summary` field from the webhook payload can be used to define the cluster on which the application is deployed.

File: internal/notifier/github_dispatch.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2020 The Flux authors
```
Comment (stefanprodan) [inline internal/notifier/github_dispatch.go:2]:
```suggestion
Copyright 2022 The Flux authors
```

File: docs/spec/v1beta1/provider.md [hunk 3]
```diff
@@ -730,3 +732,141 @@ kubectl create secret generic webhook-url \
 --from-literal=address="Endpoint=sb://fluxv2.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=yoursaskeygeneatedbyazure"
 ```
 
+### GitHub repository dispatch
+
+The `githubdispatch` webhook triggers a GitHub webhook event called [`repository_dispatch`](https://docs.github.com/en/rest/reference/repos#create-a-repository-dispatch-event) for the selected repository. The `repository_dispatch` event can be used to trigger a GitHub Actions workflow with tests for the deployed service.
```
Comment (stefanprodan) [inline docs/spec/v1beta1/provider.md:737]:
```suggestion
The `githubdispatch` provider generates GitHub  events of type [`repository_dispatch`](https://docs.github.com/en/rest/reference/repos#create-a-repository-dispatch-event) for the selected repository. The `repository_dispatch` events can be used to trigger GitHub Actions workflow.
```

File: docs/spec/v1beta1/provider.md [hunk 4]
```diff
@@ -730,3 +732,141 @@ kubectl create secret generic webhook-url \
...
+
+The request includes the `event_type` and `client_payload` fields:
+
+* The `event_type` is generated by GitHub Dispatch provider by combining the Kind, Name and Namespace of the involved object in the format `{Kind}/{Name}.{Namespace}`. For example, the `event_type` for Flux configuration named `podinfo` in the `flux-system` namespace looks like this: `Kustomization/podinfo.flux-system`
```
Comment (stefanprodan) [inline docs/spec/v1beta1/provider.md:741]:
```suggestion
* The `event_type` is generated by GitHub Dispatch provider by combining the Kind, Name and Namespace of the involved object in the format `{Kind}/{Name}.{Namespace}`. For example, the `event_type` for a Flux Kustomization named `podinfo` in the `flux-system` namespace looks like this: `Kustomization/podinfo.flux-system`.
```

File: docs/spec/v1beta1/provider.md [hunk 5]
```diff
@@ -730,3 +732,141 @@ kubectl create secret generic webhook-url \
...
+
+* The `client_payload` contains the event data from the notification controller and looks like this:
```
Comment (stefanprodan) [inline docs/spec/v1beta1/provider.md:743]:
```suggestion
* The `client_payload` contains the Kubernetes event issued by Flux, e.g.:
```

File: docs/spec/v1beta1/provider.md [hunk 6]
```diff
@@ -730,3 +732,141 @@ kubectl create secret generic webhook-url \
...
+  reportingController: kustomize-controller,
+  reportingInstance: kustomize-controller-79464d8dc5-nb9c4,
+  severity: info,
+  timestamp: 2022-04-20T12:20:28Z
+}
+```
+
+#### Setting up a GitHub workflow
+
+You need to add `repository_dispatch` trigger to trigger your GitHub workflow using the Flux notification with the relevant event type:
```
Comment (stefanprodan) [inline docs/spec/v1beta1/provider.md:770]: To trigger a GitHub Actions workflow when a Flux Kustomization finishes reconciling, you need to set the event type on the `repository_dispatch` to match the Flux object ID:

File: docs/spec/v1beta1/provider.md [hunk 7]
```diff
@@ -730,3 +732,141 @@ kubectl create secret generic webhook-url \
...
+jobs:
+  run-tests-staging:
+    if: github.event.client_payload.metadata.summary == 'staging (us-west-2)'
+    runs-on: ubuntu-18.04
+    steps:
+    - name: Run tests
+      run: echo "running tests.."
+```
+
+### Setting up the GitHub dispatch provider
```
Comment (stefanprodan) [inline docs/spec/v1beta1/provider.md:841]: Let's move this before "Setting up a GitHub workflow" please.

Comment (alekspog) [reply docs/spec/v1beta1/provider.md]: Fixed

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @alekspog 🥇

**Pull request merged**